core: make StateDirectory= or friends support '-' prefix

yuwata opened the pull request:
If a path has '-' prefix, then any errors in preparing the directory will be removed.

Closes #9296.

yuwata commented:
Note. This breaks a backward compatibility. If a unit has `StateDirectory=-foo`, then before this, /var/lib/-foo was created, but after this, /var/lib/foo will be created. Please give me any comments and suggestions.

keszybz commented on src/core/dbus-execute.c:
```diff
@@ -2087,52 +2124,53 @@ int bus_exec_context_set_transient_property(
                 STRV_FOREACH(p, l) {
-                        if (!path_is_normalized(*p))
-                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "%s= path is not normalized: %s", name, *p);
+                        char *q = *p + (**p == '-');
```
`const char *` ...

yuwata replied:
It is not necessary, but ok.

keszybz commented on src/core/dbus-execute.c:
```diff
@@ -2087,52 +2124,53 @@ int bus_exec_context_set_transient_property(
+                        } else
+                                STRV_FOREACH(p, l) {
+                                        bool ignore = false;
+                                        char *q = *p;
```
`const` here too?

yuwata replied:
It is possible, but `.path` is `char *`. Casting from `char*` to `const char*` here and revert it at the later substitution may be not good, I think.

keszybz commented on src/core/execute.c:
```diff
@@ -3963,10 +4038,15 @@ void exec_context_dump(const ExecContext *c, FILE* f, const char *prefix) {
+                        fprintf(f, "%s%s%s: %s\n", prefix,
+                                exec_directory_type_to_string(dt),
+                                c->directories[dt].directories[j].ignore ? "-" : "",
+                                c->directories[dt].directories[j].path);
```
This gives something like "RuntimeDirectory-: foobar" (the dash is in the wrong place).

yuwata replied:
ouch. fixed.

keszybz commented:
> If a unit has StateDirectory=-foo, then before this, /var/lib/-foo was created, but after this, /var/lib/foo will be created.

I think this is OK. Let's assume that nobody creates directories like that. (Even though we have `-.slice`, yuck.)

yuwata commented:
Updated. I thought that this may be useful not only for the case #9296, but also `RootImage/RootDirectory=` with read-only image or directory. But it may need more work to support such read-only systems. I will test this more in detail. So, tentatively, I have added 'dont-merge' tag.

yuwata commented:
Closing this, as it does not work at least in the current form.
